test/teos/unit/test_internal_api.py

```diff
@@ -0,0 +1,234 @@
+
+
+def test_register_wrong_user_id(internal_api, stub):
+    # If the user id is wrong we should get INVALID_ARGUMENT with the proper message
+    wrong_user_id = get_random_value_hex(32)
+
+    with pytest.raises(grpc.RpcError) as e:
+        stub.register(RegisterRequest(user_id=wrong_user_id))
+    assert e.value.code() == grpc.StatusCode.INVALID_ARGUMENT
+    assert "Provided public key does not match expected format" in e.value.details()
+
+
+def test_add_appointment(internal_api, stub):
+    # Normal request should work just fine (user needs to be registered)
+    stub.register(RegisterRequest(user_id=user_id))
+
+    appointment, _ = generate_dummy_appointment()
+    appointment_signature = Cryptographer.sign(appointment.serialize(), user_sk)
+    response = send_appointment(stub, appointment, appointment_signature)
+
+    assert isinstance(response, AddAppointmentResponse)
+
+
+def test_add_appointment_non_registered(internal_api, stub):
+    # If the user is not registered we should get UNAUTHENTICATED + the proper message
+    another_user_sk, another_user_pk = generate_keypair()
+    appointment, _ = generate_dummy_appointment()
+    appointment_signature = Cryptographer.sign(appointment.serialize(), another_user_sk)
+
+    e = send_wrong_appointment(stub, appointment, appointment_signature)
+
+    assert e.value.code() == grpc.StatusCode.UNAUTHENTICATED
+    assert "Invalid signature or user does not have enough slots available" in e.value.details()
+
+
+def test_add_appointment_not_enough_slots(internal_api, stub):
+    # UNAUTHENTICATED should also be get if the user does not have enough appointment slots
+    # Register the user and set the slots to 0
+    stub.register(RegisterRequest(user_id=user_id))
+    internal_api.watcher.gatekeeper.registered_users[user_id].available_slots = 0
+
+    appointment, _ = generate_dummy_appointment()
+    appointment_signature = Cryptographer.sign(appointment.serialize(), user_sk)
+
+    e = send_wrong_appointment(stub, appointment, appointment_signature)
+
+    assert e.value.code() == grpc.StatusCode.UNAUTHENTICATED
+    assert "Invalid signature or user does not have enough slots available" in e.value.details()
+
+
+def test_add_appointment_limit_reached(internal_api, stub):
+    # If the tower appointment limit is reached RESOURCE_EXHAUSTED should be returned
+    current_limit = internal_api.watcher.max_appointments
+    internal_api.watcher.max_appointments = 0
+
+    stub.register(RegisterRequest(user_id=user_id))
+
+    appointment, _ = generate_dummy_appointment()
+    appointment_signature = Cryptographer.sign(appointment.serialize(), user_sk)
+
+    e = send_wrong_appointment(stub, appointment, appointment_signature)
+
+    assert e.value.code() == grpc.StatusCode.RESOURCE_EXHAUSTED
+    assert "Appointment limit reached" in e.value.details()
+
+    # Set the limit back just in case the whole suite if being run
+    internal_api.watcher.max_appointments = current_limit
```

Comment:
Instead of changing the `internal_api` object (which might still create problems if an assertion fails, for example), you can use the `monkeypatch` fixture:

```python
def test_add_appointment_limit_reached(internal_api, stub, monkeypatch):
    # If the tower appointment limit is reached RESOURCE_EXHAUSTED should be returned
    monkeypatch.setattr(internal_api.watcher, "max_appointments", 0)

    stub.register(RegisterRequest(user_id=user_id))

    appointment, _ = generate_dummy_appointment()
    appointment_signature = Cryptographer.sign(appointment.serialize(), user_sk)

    e = send_wrong_appointment(stub, appointment, appointment_signature)

    assert e.value.code() == grpc.StatusCode.RESOURCE_EXHAUSTED
    assert "Appointment limit reached" in e.value.details()
```

Reply:
Done in 563417d